hazelcast/src/main/java/com/hazelcast/util/executor/LoggingScheduledExecutor.java

```diff
@@ -65,7 +88,8 @@ public LoggingScheduledExecutor(ILogger logger, int corePoolSize, ThreadFactory
                                     boolean removeOnCancel) {
         super(corePoolSize, threadFactory);
         this.logger = checkNotNull(logger, "logger cannot be null");
-        this.removeOnCancel = removeOnCancel;
+        this.wrap = !trySetRemoveOnCancelPolicy() && removeOnCancel;
```

Comment:
This looks a bit brittle. It's a condition check with side effects and will behave differently if the order is changed to `removeOnCancel && !trySetRemoveOnCancelPolicy()`.

My question is, if it's wanted that `setRemoveOnCancelPolicy(true)` is always executed, regardless of the `removeOnCancel` flag. If so, I would move the `removeOnCancel` handling into the `trySetRemoveOnCancelPolicy`:
```java
private boolean trySetRemoveOnCancelPolicy() {
    try {
        Method method = ScheduledThreadPoolExecutor.class.getMethod("setRemoveOnCancelPolicy", Boolean.TYPE);
        method.invoke(this, true);
        return removeOnCancel;
    } catch (Exception e) {
        removeOnCancel = false;
        return false;
    }
}
```
Now you can call it just with `this.wrap = trySetRemoveOnCancelPolicy();` and don't have to overwrite `removeOnCancel` twice in the code.

Reply:
I made the logic a bit less compact:

```
  if (trySetRemoveOnCancelPolicy()) {
            manualRemoveOnCancel = false;
        } else {
            manualRemoveOnCancel = removeOnCancel;
        }
```

Reply:
Yes. The setRemoveOnCancelPolicy(true) should always be called. The removal of a task is done in constant time in Java 7

```
  private int indexOf(Object x) {
            if (x != null) {
                if (x instanceof ScheduledFutureTask) {<---- constant time lookup
                    int i = ((ScheduledFutureTask) x).heapIndex;
                    // Sanity check; x could conceivably be a
                    // ScheduledFutureTask from some other pool.
                    if (i >= 0 && i < size && queue[i] == x)
                        return i;
                } else {
                    for (int i = 0; i < size; i++)
                        if (x.equals(queue[i]))
                            return i;
                }
            }
            return -1;
        }

```

So if the automatic removal of canceled futures is possible, it should be done. Not doing it only has nasty side effects like retaining memory.